File: 429. N-ary Tree Level Order Traversal/main.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    // bfs method
    // public List<List<Integer>> levelOrder(Node root) {
    // if (root==null){
    // return new LinkedList();
    // }
    // List<Node> stack = new ArrayList<Node>();
    // stack.add(root);
    // List<List<Integer>> result = new LinkedList();
    // while (!stack.isEmpty()){
    // List<Integer> cls = new ArrayList<>(); // cls: current level stack
    // List<Node> tmpStack = new ArrayList<>();
    // for (Node node:stack){
    // cls.add(node.val);
    // if (node.children!=null){
    // for (Node child:node.children){
    // tmpStack.add(child);
    // }
    // }
    // }
    // stack = tmpStack;
    // result.add(cls);
    // }
    // return result;
    // }

    // solution with dfs
    public List<List<Integer>> levelOrder(Node root) {
        List<List<Integer>> count = new ArrayList<>();
        dfs(root, 0, count);
        return count;
    }

    public void dfs(Node root, int i, List<List<Integer>> count) {
        if (root == null)
            return;
        if (i > count.size() - 1) {
            count.add(new ArrayList<Integer>());
            count.get(i).add(root.val);
        } else {
            count.get(i).add(root.val);
        }
        if (root.children != null) {
            for (Node child : root.children) {
                dfs(child, i + 1, count);
            }
        }
    }
}

File: 429. N-ary Tree Level Order Traversal/main.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""


class Solution:

    def levelOrder(self, root: 'Node') -> List[List[int]]:
        if not root:
            return None

        # bfs
        # from collections import deque
        # dq = deque()
        # dq.append(root)
        # result = []
        # while dq:
        #     cll = len(dq)  # current level length
        #     tmp_list = []
        #     for i in range(cll):
        #         node = dq.popleft()
        #         tmp_list.append(node.val)
        #         if node.children:
        #             for child in node.children:
        #                 dq.append(child)
        #     result.append(tmp_list)
        # return result

        # opTimized code
        stack, result = [root], []
        while stack:
            result.append([node.val for node in stack])
            stack = [child for node in stack for child in node.children if child]
        return result
